File: 912_Sort_an_Array/quickSort.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        int len = nums.size();
        sortArray(nums, 0, len - 1);
        return nums;
    }

    void sortArray(vector<int> &nums, int lo, int hi) {
        if (hi <= lo) return;
        int i = rand() % (hi - lo + 1) + lo;
        swap(nums[lo], nums[i]);
        int p = partition(nums, lo, hi);
        sortArray(nums, lo, p - 1);
        sortArray(nums, p + 1, hi);
    }

    int partition(vector<int> &nums, int lo, int hi) {
        int v = nums[lo];
        int i = lo + 1, j = hi;
        while (true) {
            while (nums[i] <= v) {
                i++;
                if (i > hi) break;
            }
            while (nums[j] > v) {
                j--;
                if (j == lo) break;
            }
            if (i >= j) break;
            swap(nums[i], nums[j]);
        }
        nums[lo] = nums[j];
        nums[j] = v;
        return j;
    }
};

int main() {
    Solution s;
    vector<int> nums {5,2,3,1,-1,-34,45};
    for (int i : s.sortArray(nums)) {
        cout << i << " ";
    }
    cout << "\n";
}

File: 912_Sort_an_Array/quickSort.java
import java.util.Random;

class Solution {
    private int[] aux;
    public int[] sortArray(int[] nums) {
        int n = nums.length;
        Random r = new Random();
        int i = r.nextInt(n);
        swap(nums, 0, i);
        quickSort(nums, 0, n - 1);
        return nums;
    }

    private void quickSort(int[] nums, int lo, int hi) {
        if (lo >= hi) return;
        int last = lo, i;
        for (i = last + 1; i <= hi; i++) {
            if (nums[i] < nums[lo]) {
                swap(nums, ++last, i);
            }
        }
        swap(nums, lo, last);

        quickSort(nums, lo, last - 1);
        quickSort(nums, last + 1, hi);
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}

public class quickSort {
    public static void main(String[] args) {
        Solution s = new Solution();
        int[] nums = {5,2,3,1,-1,-34,45};
        s.sortArray(nums);
        for (int i : nums) 
            System.out.print(i + " ");
        System.out.println();
    }
}
